metadata-ingestion/src/datahub/ingestion/source/cassandra.py

```diff
@@ -0,0 +1,592 @@
+@capability(SourceCapability.PLATFORM_INSTANCE, "Enabled by default")
+class CassandraSource(Source):
+
+    """
+    This plugin extracts the following:
+
+    - Metadata for tables
+    - Column types associated with each table column
+    - The keyspace each table belongs to
+    """
+
+    config: CassandraSourceConfig
+    report: CassandraSourceReport
+    cassandra_session: Session
+    platform: str
+
+    def __init__(self, ctx: PipelineContext, config: CassandraSourceConfig):
+        self.ctx = ctx
+        self.platform = PLATFORM_NAME_IN_DATAHUB
+        self.config = config
+        self.report = CassandraSourceReport()
+
+        # attempt to connect to cass
+        auth_provider: AuthProvider = None
+        if config.username and config.password:
+            ssl_context = SSLContext(PROTOCOL_TLSv1_2)
+            ssl_context.verify_mode = CERT_NONE
+            auth_provider = PlainTextAuthProvider(
+                username=config.username, password=config.password
+            )
+
+        cluster: Cluster = (
+            Cluster(
+                [config.contact_point],
+                port=config.port,
+                auth_provider=auth_provider,
+                ssl_context=ssl_context,
+            )
+            if auth_provider
+            else Cluster(
+                [config.contact_point],
+                port=config.port,
+            )
+        )
+        session: Session = cluster.connect()
+        self.cassandra_session = session
+
+    @classmethod
+    def create(cls, config_dict, ctx):
+        config = CassandraSourceConfig.parse_obj(config_dict)
+        return cls(ctx, config)
+
+    def get_workunits_internal(
+        self,
+    ) -> Iterable[MetadataWorkUnit]:
+
+        # get all keyspaces and iterate through them
+        # TODO: create a type for the response
+        keyspaces = self.cassandra_session.execute(GET_KEYSPACES_QUERY)
+        keyspaces = sorted(
+            keyspaces,
+            key=lambda k: getattr(
+                k, CASSANDRA_SYSTEM_SCHEMA_COLUMN_NAMES["keyspace_name"]
+            ),
+        )
+
+        for keyspace in keyspaces:
+            keyspace_name: str = getattr(
+                keyspace, CASSANDRA_SYSTEM_SCHEMA_COLUMN_NAMES["keyspace_name"]
+            )
+
+            # skip system keyspaces
+            if keyspace_name in SYSTEM_KEYSPACE_LIST:
+                continue
+            # skip keyspaces not allowed by the config
+            if not self.config.keyspace_pattern.allowed(keyspace_name):
+                self.report.report_dropped(keyspace_name)
+                continue
+
+            # 1. Construct and emit the container aspect.
+            yield from self._generate_keyspace_container(keyspace_name)
+
+            # 2. get all tables for this keyspace and emit their metadata (dataset, container and schema info)
+            try:
+                yield from self._extract_tables_from_keyspace(keyspace_name)
+            except Exception as e:
```

Comment:
some other sources don't have error handling, should I remove this?
ie. does a hard failure help flag an issue and get people's attention vs soft reporting?